url063b9533f6_karlney_tetris_ai_tgz-pJ8-karlney_tetris_cmd_TetrisGameRunnerJ8/src/karlney/tetris/ai/BoardMeasuresUtil.java

```diff
@@ -0,0 +1,233 @@
+        for(int i= cols; i>0; i--){
+            if ( board[i][j].isFilled()){
+                nr++;
+            }
+        }
+        return nr;
+    }
+
+    public int getMaxHeight(){
+        for(int j=0; j< rows; j++){
+            for(int i= cols; i>0; i--){
+                if ( board[i][j].isFilled()){
+                    return rows -j;
+                }
+            }
+        }
+        return 0;
+    }
+
+
+    public int getNrOfBlocks(){
+        int nr=0;
+        for(int j= rows; j>0; j--){
+            for(int i= cols; i>0; i--){
+                if ( board[i][j].isFilled()){
+                    nr++;
+                }
+            }
+        }
+        return nr;
+    }
+
+    public double checkForSequences(){
+        double out=0;
+        for(int j= rows; j>0; j--){
+            int seq=0;
+            for(int i= cols; i>0; i--){
+                if ( board[i][j].isFilled()){
+                    seq++;
+                }else{
+                    out+= Math.pow(2,seq)-1;
+                }
+            }
+        }
+        return out;
+    }
+
+
+    /*
+1. Landing Height: The height where the piece is put (= the height of the column + (the height of the piece / 2))
+
+  var lh = last_move.landing_height + ((last_move.piece.length - 1) / 2);
+
+     */
+    public static double getLandingHeight(Board board, Piece piece) {
+        return getColumnHeight(board,piece)+((getPieceHeight(piece)-1)/2.0);
+    }
+
+    private static int getEmptyTopRows(Piece piece) {
+        boolean filled = false;
+        for (int i=0; i<piece.getSize();i++){
+            filled = filled || piece.isFilled(i,0);
+        }
+        return filled?0:1;
+    }
+
+    private static double getColumnHeight(Board board,Piece piece) {
+        int y = piece.getY()+getEmptyTopRows(piece);
+        return board.getRows()-y-getPieceHeight(piece)+1;
+    }
+
+    private static double getPieceHeight(Piece piece) {
+        int start = -1;
+        for (int j=0;j<piece.getSize();j++){
+            for (int i=0; i<piece.getSize();i++){
+                if (start<0 && piece.isFilled(i,j)){
+                    start=j;
+                }
+            }
+        }
+        int end = -1;
+        for (int j=piece.getSize();j>0;j--){
+            for (int i=0; i<piece.getSize();i++){
+                if (end<0 && piece.isFilled(i,j-1)){
+                    end=j-1;
+                }
+            }
+        }
+        return end-start+1;
+    }
+
+
+    /*
+3. Row Transitions: The total number of row transitions. A row transition occurs when an empty cell is adjacent to a filled cell on the same row and vice versa.
+    */
+    public static int getRowTransitions(Board board) {
+        int nr=0;
+        for (int j=board.getRows(); j>0; j--){
+            for (int i=0; i<board.getCols()+1; i++)	{
+                if (board.isFilled(i,j)!=board.isFilled(i+1,j)){
+                    nr++;
+                }
+            }
+        }
+        return nr;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 141:149, 156:164.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.